File: game/resources/level.lua
level = {
    spawn_position = { -- The position of the spawn.
        x = 120,
        y = 80,
    },
    max_player = 1, --The maximum number of players of the level
    players = { --The player entities of the level (the first one will be the player 1, ...), need at least as many as the max_player count.
        "kenney_player1",
    },
    objects = { --The entities template instances with their parameter fulfilled
        {
            id = 1,
            template = "kenney_block_grass",
            values = {
                x = 1*64,
                y = 6*64,
            }
        },
        {
            id = 2,
            template = "kenney_block_grass",
            values = {
                x = 2*64,
                y = 6*64,
            }
        },
        {
            id = 3,
            template = "kenney_block_grass",
            values = {
                x = 3*64,
                y = 6*64,
            }
        },
        {
            id = 4,
            template = "kenney_block_grass",
            values = {
                x = 4*64,
                y = 6*64,
            }
        },
        {
            id = 5,
            template = "kenney_block_grass_center",
            values = {
                x = 1*64,
                y = 7*64,
            }
        },
        {
            id = 6,
            template = "kenney_block_grass_center",
            values = {
                x = 2*64,
                y = 7*64,
            }
        },
        {
            id = 7,
            template = "kenney_block_grass_center",
            values = {
                x = 3*64,
                y = 7*64,
            }
        },
        {
            id = 8,
            template = "kenney_block_grass_center",
            values = {
                x = 4*64,
                y = 7*64,
            }
        },
        {
            id = 9,
            template = "kenney_block_wood_platform",
            values = {
                x = 2*64,
                y = 4.5*64,
            }
        },
        {
            id = 10,
            template = "kenney_block_wood_platform",
            values = {
                x = 3*64,
                y = 4.5*64,
            }
        },
        {
            id = 11,
            template = "kenney_block_grass_center",
            values = {
                x = 1*64,
                y = 11*64,
            }
        },
        {
            id = 12,
            template = "kenney_block_grass_center",
            values = {
                x = 2*64,
                y = 11*64,
            }
        },
        {
            id = 13,
            template = "kenney_block_grass_center",
            values = {
                x = 3*64,
                y = 11*64,
            }
        },
        {
            id = 14,
            template = "kenney_block_grass_center",
            values = {
                x = 4*64,
                y = 11*64,
            }
        },
        {
            id = 15,
            template = "kenney_block_grass_center",
            values = {
                x = 5*64,
                y = 11*64,
            }
        },
        {
            id = 16,
            template = "kenney_block_grass_center",
            values = {
                x = 6*64,
                y = 11*64,
            }
        },
        {
            id = 17,
            template = "kenney_block_grass_center",
            values = {
                x = 7*64,
                y = 11*64,
            }
        },
        {
            id = 18,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 11*64,
            }
        },
        {
            id = 19,
            template = "kenney_block_grass_center",
            values = {
                x = 9*64,
                y = 11*64,
            }
        },
        {
            id = 20,
            template = "kenney_block_grass_center",
            values = {
                x = 10*64,
                y = 11*64,
            }
        },
        {
            id = 21,
            template = "kenney_block_grass_center",
            values = {
                x = 11*64,
                y = 11*64,
            }
        },
        {
            id = 22,
            template = "kenney_block_grass_center",
            values = {
                x = 12*64,
                y = 11*64,
            }
        },
        {
            id = 23,
            template = "kenney_block_grass_center",
            values = {
                x = 1*64,
                y = 10*64,
            }
        },
        {
            id = 24,
            template = "kenney_block_grass_center",
            values = {
                x = 1*64,
                y = 9*64,
            }
        },
        {
            id = 25,
            template = "kenney_block_grass_center",
            values = {
                x = 1*64,
                y = 8*64,
            }
        },
        {
            id = 26,
            template = "kenney_block_grass_center",
            values = {
                x = 13*64,
                y = 11*64,
            }
        },
        {
            id = 27,
            template = "kenney_block_grass_center",
            values = {
                x = 14*64,
                y = 11*64,
            }
        },
        {
            id = 28,
            template = "kenney_block_grass_center",
            values = {
                x = 14*64,
                y = 10*64,
            }
        },
        {
            id = 29,
            template = "kenney_block_grass_center",
            values = {
                x = 14*64,
                y = 9*64,
            }
        },
        {
            id = 30,
            template = "kenney_block_grass_center",
            values = {
                x = 14*64,
                y = 8*64,
            }
        },
        {
            id = 31,
            template = "kenney_block_wood_platform",
            values = {
                x = 5*64,
                y = 6*64,
            }
        },
        {
            id = 31,
            template = "kenney_block_wood_platform",
            values = {
                x = 7*64,
                y = 6*64,
            }
        },
        {
            id = 31,
            template = "kenney_block_wood_platform",
            values = {
                x = 6*64,
                y = 7*64,
            }
        },
        {
            id = 32,
            template = "kenney_block_wood_platform",
            values = {
                x = 5*64,
                y = 10*64,
            }
        },
        {
            id = 35,
            template = "kenney_block_wood_platform",
            values = {
                x = 6*64,
                y = 9*64,
            }
        },
        {
            id = 36,
            template = "kenney_block_wood_platform",
            values = {
                x = 7*64,
                y = 9*64,
            }
        },
        {
            id = 37,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 8*64,
            }
        },
        {
            id = 38,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 7*64,
            }
        },
        {
            id = 39,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 6*64,
            }
        },
        {
            id = 40,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 5*64,
            }
        },
        {
            id = 41,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 4*64,
            }
        },
        {
            id = 42,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 3*64,
            }
        },
        {
            id = 43,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 2*64,
            }
        },
        {
            id = 44,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 1*64,
            }
        },
        {
            id = 45,
            template = "kenney_block_grass_center",
            values = {
                x = 8*64,
                y = 0*64,
            }
        },
        {
            id = 46,
            template = "kenney_block_grass_center",
            values = {
                x = 14*64,
                y = 7*64,
            }
        },
        {
            id = 47,
            template = "kenney_block_grass",
            values = {
                x = 14*64,
                y = 6*64,
            }
        },
        {
            id = 48,
            template = "kenney_block_grass",
            values = {
                x = 10*64,
                y = 10*64,
            }
        },
        {
            id = 49,
            template = "kenney_block_grass_slope_right",
            values = {
                x = 11*64,
                y = 9*64,
            }
        },
        {
            id = 50,
            template = "kenney_block_grass_slope_right",
            values = {
                x = 12*64,
                y = 8*64,
            }
        },
        {
            id = 51,
            template = "kenney_block_grass_slope_right",
            values = {
                x = 13*64,
                y = 7*64,
            }
        },
        {
            id = 52,
            template = "kenney_block_grass_slope_right_center",
            values = {
                x = 11*64,
                y = 10*64,
            }
        },
        {
            id = 53,
            template = "kenney_block_grass_slope_right_center",
            values = {
                x = 12*64,
                y = 9*64,
            }
        },
        {
            id = 53,
            template = "kenney_block_grass_slope_right_center",
            values = {
                x = 13*64,
                y = 8*64,
            }
        },
        {
            id = 54,
            template = "kenney_block_grass_center",
            values = {
                x = 13*64,
                y = 9*64,
            }
        },
        {
            id = 55,
            template = "kenney_block_grass_center",
            values = {
                x = 13*64,
                y = 10*64,
            }
        },
        {
            id = 56,
            template = "kenney_block_grass_center",
            values = {
                x = 12*64,
                y = 10*64,
            }
        },
        {
            id = 57,
            template = "kenney_block_grass_slope_right",
            values = {
                x = 9*64,
                y = 10*64,
            }
        },
        {
            id = 58,
            template = "kenney_block_wood_platform",
            values = {
                x = 9*64,
                y = 4.5*64,
            }
        },
        {
            id = 59,
            template = "kenney_block_wood_platform",
            values = {
                x = 10*64,
                y = 4.5*64,
            }
        },
        {
            id = 60,
            template = "kenney_block_wood_platform",
            values = {
                x = 11*64,
                y = 4.5*64,
            }
        },
        {
            id = 61,
            template = "kenney_block_wood_platform",
            values = {
                x = 12*64,
                y = 4.5*64,
            }
        },
    }
}

File: game/src/Level/Level.cpp
#include "Level/Level.hpp"

#include <exception>

#include "Components/Component.hpp"
#include "Components/PlayerComponent.hpp"
#include "Components/TemplateComponent.hpp"
#include "Settings/tinyxml2.h"

namespace level
{

Level::Level(lua::LuaState& luaState, LevelMode levelMode) :
    m_luaState(luaState),
    m_eventMgr(),
    m_entityMgr(m_eventMgr),
    m_spawnPosition(),
    m_playersTemplates(),
    m_levelMode(levelMode),
    m_nextId(0)
{

}

void Level::LoadFromFile(const std::string& path)
{
    for(entityx::Entity entity : m_entityMgr.entities_for_debugging())
        entity.destroy();
    m_playersTemplates.clear();
    m_nextId = 0;

    std::cout << "Loading level \"" << path << "\"..." << std::endl;

    tinyxml2::XMLDocument levelDocument;
    if(levelDocument.LoadFile(path.c_str()) != tinyxml2::XML_NO_ERROR)
    {
        throw std::runtime_error("[Level/Error] Can't load the level \"" + path + "\" ! Not found or parsing error !");
    }

    tinyxml2::XMLElement* levelElement = levelDocument.RootElement();
    if(!levelElement)
    {
        throw std::runtime_error("[Level/Error] Can't find <level> root markup in the level !");
    }

    //Load the spawn position
    tinyxml2::XMLElement* spawnPosElement = levelElement->FirstChildElement("spawn_position");
    if(!spawnPosElement)
    {
        throw std::runtime_error("[Level/Error] Can't find <spawn_position> markup in the level !");
    }
    if(spawnPosElement->QueryFloatAttribute("x", &(m_spawnPosition.x)) != tinyxml2::XML_NO_ERROR || spawnPosElement->QueryFloatAttribute("y", &(m_spawnPosition.y)) != tinyxml2::XML_NO_ERROR)
    {
        throw std::runtime_error("[Level/Error] Can't find x and y in <spawn_position> markup in the level !");
    }

    //Load the players templates names
    tinyxml2::XMLElement* playersElement = levelElement->FirstChildElement("players");
    if(!playersElement)
    {
        throw std::runtime_error("[Level/Error] Can't find <players> markup in the level !");
    }
    for(tinyxml2::XMLElement* playerElem = playersElement->FirstChildElement("player");
        playerElem != nullptr;
        playerElem = playersElement->NextSiblingElement("player"))
    {
        const char* playerTemplate = playerElem->Attribute("entity_template");
        if(!playerTemplate)
        {
            std::cout << "[Level/Warning] Invalid player template !" << std::endl;
            continue;
        }
        m_playersTemplates.push_back(std::string(playerTemplate));
    }

    //First, load all the entity and register them to the SerializedEntityGetter
    tinyxml2::XMLElement* objectsElement = levelElement->FirstChildElement("objects");
    if(!objectsElement)
    {
        throw std::runtime_error("[Level/Error] Can't find <object> markup in the level !");
    }

    SerializedEntityGetter entityGetter;
    std::vector<entityx::Entity> createdEntities;

    for(tinyxml2::XMLElement* objectElem = objectsElement->FirstChildElement("object");
        objectElem != nullptr;
        objectElem = objectElem->NextSiblingElement("object"))
    {
        //Create the entity
        entityx::Entity newEntity = m_entityMgr.create();
        createdEntities.push_back(newEntity);

        //Register it with its id (if it has a id attribute)
        int newEntityId = -1;
        auto error = objectElem->QueryIntAttribute("id", &newEntityId);
        if(error == tinyxml2::XML_NO_ERROR)
        {
            entityGetter.registerEntity(newEntity, newEntityId);
            m_nextId = std::max(m_nextId, newEntityId);
        }
    }

    //Now that all entities are created and registered, iterate all of them to
    //assign the components according to their template !
    unsigned int i = 0;
    for(tinyxml2::XMLElement* objectElem = objectsElement->FirstChildElement("object");
        objectElem != nullptr;
        objectElem = objectElem->NextSiblingElement("object"))
    {
        if(!objectElem->Attribute("template"))
        {
            //There's no template defined, remove the entity.
            std::cout << "[Level/Warning] No template attribute defined for an <object> !" << std::endl;
            createdEntities.erase(createdEntities.begin() + i);
            continue;
        }
        std::string entityTemplateName(objectElem->Attribute("template"));

        tinyxml2::XMLElement* entityParameters = objectElem->FirstChildElement("parameters");
        if(!entityParameters)
        {
            //There's no template defined, remove the entity.
            std::cout << "[Level/Warning] No <parameters> in <object> !" << std::endl;
            createdEntities.erase(createdEntities.begin() + i);
            continue;
        }

        m_luaState.getTemplate(entityTemplateName).initializeEntity(createdEntities[i], entityGetter, entityParameters, /*templateComponent*/ m_levelMode == LevelMode::EditMode);

        ++i;
    }

    //Put the current level instance into "current_level" lua global variable
    m_luaState.getState().set("current_level", this);

    std::cout << "Level successfully loaded (" << createdEntities.size() << " entities)." << std::endl;

    //TODO: Support multiple players creation
    if(m_levelMode == LevelMode::PlayMode)
    {
        std::cout << "Creating players..." << std::endl;

        const lua::EntityTemplate& playerTemplate = m_luaState.getTemplate(m_playersTemplates[0]);

        entityx::Entity playerEntity = m_entityMgr.create();
        playerTemplate.initializeEntity(playerEntity, entityGetter);

        //Set x and y parameters according to spawn_position
        auto& parameters = playerTemplate.getParameters();

        const lua::EntityTemplate::Parameter& xParameter = parameters.at("x");
        const lua::EntityTemplate::Parameter& yParameter = parameters.at("y");

        lua::EntityHandle(playerEntity).setAttributeAsAny(xParameter.component, xParameter.attribute, m_spawnPosition.x);
        lua::EntityHandle(playerEntity).setAttributeAsAny(yParameter.component, yParameter.attribute, m_spawnPosition.y);

        if(!playerEntity.has_component<components::PlayerComponent>())
            throw std::runtime_error(std::string("[Level/Error] Player entities must have the \"Player\" component declared in their template ! Not the case with \"") + m_playersTemplates[0] + std::string("\""));
        playerEntity.component<components::PlayerComponent>()->playerNumber = 0;

        std::cout << "Players created." << std::endl;
    }
    /////////////////////////////////////////
}

void Level::SaveToFile(const std::string& path)
{
    if(m_levelMode == LevelMode::PlayMode)
    {
        throw std::runtime_error("[Level/Error] Can't save a level opened in play mode !");
    }

    tinyxml2::XMLDocument levelDocument;

    tinyxml2::XMLElement* levelElement = levelDocument.NewElement("level");
    levelDocument.LinkEndChild(levelElement);

    //Save the spawn position
    {
        tinyxml2::XMLElement* spawnElement = levelDocument.NewElement("spawn_position");
        spawnElement->SetAttribute("x", m_spawnPosition.x);
        spawnElement->SetAttribute("y", m_spawnPosition.y);

        levelElement->LinkEndChild(spawnElement);
    }

    //Save the players templates
    {
        tinyxml2::XMLElement* playersElement = levelDocument.NewElement("players");
        playersElement->SetAttribute("max_player", static_cast<int>(m_playersTemplates.size()));

        for(auto it = m_playersTemplates.cbegin(); it != m_playersTemplates.cend(); ++it)
        {
            tinyxml2::XMLElement* playerElement = levelDocument.NewElement("player");
            playerElement->SetAttribute("entity_template", (*it).c_str());

            playersElement->LinkEndChild(playerElement);
        }

        levelElement->LinkEndChild(playersElement);
    }

    //Save the entities
    {
        tinyxml2::XMLElement* objectsElement = levelDocument.NewElement("objects");

        //Go through all entities and create an XMLElement element for them and set an ID
        //Give the ID to the SerializedEntityGetter
        std::map<entityx::Entity, tinyxml2::XMLElement*> entityElements;

        SerializedEntityGetter entityGetter;
        unsigned int nextId = 1;
        m_entityMgr.each<components::TemplateComponent>([&entityGetter, &nextId, &entityElements, &objectsElement, &levelDocument](entityx::Entity entity, components::TemplateComponent& templateComponent)
        {
            entityGetter.registerEntity(entity, nextId);

            tinyxml2::XMLElement* entityElement = levelDocument.NewElement("object");
            entityElement->SetAttribute("id", static_cast<int>(nextId));
            entityElement->SetAttribute("template", templateComponent.templateName.c_str());

            entityElements[entity] = entityElement;
            objectsElement->LinkEndChild(entityElement);

            ++nextId;
        });

        //Go through all entities again to save their parameters and entity template name
        //This allows all entities to be able to get the other's ID while saving their parameters
        m_entityMgr.each<components::TemplateComponent>([&](entityx::Entity entity, components::TemplateComponent& templateComponent)
        {
            const lua::EntityTemplate& entityTemplate = m_luaState.getTemplate(templateComponent.templateName);

            tinyxml2::XMLElement* parametersElement = levelDocument.NewElement("parameters");
            entityElements.at(entity)->LinkEndChild(parametersElement);

            entityTemplate.saveEntity(entity, entityGetter, parametersElement);
        });

        levelElement->LinkEndChild(objectsElement);
    }

    levelDocument.SaveFile(path.c_str());
}

entityx::Entity Level::createNewEntity(const std::string& templateName, bool templateComponent)
{
    entityx::Entity newEntity = m_entityMgr.create();
    m_luaState.getTemplate(templateName).initializeEntity(newEntity, SerializedEntityGetter(), templateComponent);

    if(templateComponent)
    {
        newEntity.component<components::TemplateComponent>()->serializedId = m_nextId;
        ++m_nextId;
    }

    return newEntity;
}

lua::EntityHandle Level::createNewEntityLua(const std::string& templateName)
{
    return lua::EntityHandle(createNewEntity(templateName));
}

void Level::registerClass(lua::LuaState& luaState)
{
    sol::constructors<> ctor;
    sol::usertype<Level> levelLuaClass(ctor,
        "create_new_entity", &Level::createNewEntityLua
    );
    luaState.getState().set_usertype("level", levelLuaClass);
}

}
